File: app/src/main/java/piuk/blockchain/android/ui/auth/LandingActivity.java
package piuk.blockchain.android.ui.auth;

import android.content.Context;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.v7.app.AlertDialog;
import android.view.MotionEvent;
import android.view.View;

import piuk.blockchain.android.R;
import piuk.blockchain.android.data.api.EnvironmentSettings;
import piuk.blockchain.android.data.connectivity.ConnectivityStatus;
import piuk.blockchain.android.databinding.ActivityLandingBinding;
import piuk.blockchain.android.ui.base.BaseAuthActivity;
import piuk.blockchain.android.ui.createwallet.CreateWalletActivity;
import piuk.blockchain.android.ui.customviews.ToastCustom;
import piuk.blockchain.android.ui.login.LoginActivity;
import piuk.blockchain.android.ui.recover.RecoverFundsActivity;
import piuk.blockchain.android.util.AppUtil;
import piuk.blockchain.android.util.PrefsUtil;


public class LandingActivity extends BaseAuthActivity {

    private AppUtil appUtil;

    public static void start(Context context) {
        Intent intent = new Intent(context, LandingActivity.class);
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_NEW_TASK);
        context.startActivity(intent);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        final ActivityLandingBinding binding =
                DataBindingUtil.setContentView(this, R.layout.activity_landing);
        appUtil = new AppUtil(this);

        EnvironmentSettings environmentSettings = new EnvironmentSettings();

        if (environmentSettings.shouldShowDebugMenu()) {
            ToastCustom.makeText(
                    this,
                    "Current environment: "
                            + environmentSettings.getEnvironment().getName(),
                    ToastCustom.LENGTH_SHORT,
                    ToastCustom.TYPE_GENERAL);

            binding.buttonSettings.setVisibility(View.VISIBLE);
            binding.buttonSettings.setOnClickListener(view ->
                    new EnvironmentSwitcher(this, new PrefsUtil(this)).showDebugMenu());
        }

        binding.create.setOnClickListener(view -> startCreateActivity());
        binding.login.setOnClickListener(view -> startLoginActivity());
        binding.recoverFunds.setOnClickListener(view -> showFundRecoveryWarning());

        if (!ConnectivityStatus.hasConnectivity(this)) {
            new AlertDialog.Builder(this, R.style.AlertDialogStyle)
                    .setMessage(getString(R.string.check_connectivity_exit))
                    .setCancelable(false)
                    .setPositiveButton(R.string.dialog_continue, (d, id) -> {
                        Intent intent = new Intent(LandingActivity.this, LandingActivity.class);
                        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_NEW_TASK);
                        startActivity(intent);
                    })
                    .create()
                    .show();
        }
    }

    private void startLoginActivity() {
        Intent intent = new Intent(this, LoginActivity.class);
        startActivity(intent);
    }

    private void startCreateActivity() {
        Intent intent = new Intent(this, CreateWalletActivity.class);
        startActivity(intent);
    }

    private void startRecoveryActivityFlow() {
        Intent intent = new Intent(this, RecoverFundsActivity.class);
        startActivity(intent);
    }

    private void showFundRecoveryWarning() {
        new AlertDialog.Builder(this, R.style.AlertDialogStyle)
                .setTitle(R.string.app_name)
                .setMessage(R.string.recover_funds_warning_message)
                .setPositiveButton(R.string.dialog_continue, (dialogInterface, i) -> startRecoveryActivityFlow())
                .setNegativeButton(android.R.string.cancel, null)
                .create()
                .show();
    }

    @Override
    public void startLogoutTimer() {
        // No-op
    }

    @Override
    public boolean dispatchTouchEvent(MotionEvent event) {
        // Test for screen overlays before user creates a new wallet or enters confidential information
        return appUtil.detectObscuredWindow(this, event) || super.dispatchTouchEvent(event);
    }
}

File: app/src/main/java/piuk/blockchain/android/ui/auth/LandingActivity.kt
package piuk.blockchain.android.ui.auth

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.support.v7.app.AlertDialog
import android.view.MotionEvent
import kotlinx.android.synthetic.main.activity_landing.*
import piuk.blockchain.android.R
import piuk.blockchain.android.data.connectivity.ConnectivityStatus
import piuk.blockchain.android.injection.Injector
import piuk.blockchain.android.ui.base.BaseMvpActivity
import piuk.blockchain.android.ui.createwallet.CreateWalletActivity
import piuk.blockchain.android.ui.login.LoginActivity
import piuk.blockchain.android.ui.recover.RecoverFundsActivity
import piuk.blockchain.android.util.PrefsUtil
import piuk.blockchain.android.util.extensions.toast
import piuk.blockchain.android.util.extensions.visible
import javax.inject.Inject

class LandingActivity : BaseMvpActivity<LandingView, LandingPresenter>(), LandingView {

    @Inject lateinit var landingPresenter: LandingPresenter

    init {
        Injector.getInstance().presenterComponent.inject(this)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_landing)

        create.setOnClickListener { startCreateActivity() }
        login.setOnClickListener { startLoginActivity() }
        recoverFunds.setOnClickListener { showFundRecoveryWarning() }

        if (!ConnectivityStatus.hasConnectivity(this)) {
            showConnectivityWarning()
        } else {
            presenter.initPreLoginPrompts(this)
        }

        onViewReady()
    }

    override fun createPresenter() = landingPresenter

    override fun getView() = this

    override fun startLogoutTimer() {
        // No-op
    }

    override fun dispatchTouchEvent(event: MotionEvent): Boolean {
        // Test for screen overlays before user creates a new wallet or enters confidential information
        return presenter.getAppUtil().detectObscuredWindow(this, event) || super.dispatchTouchEvent(event)
    }

    override fun showDebugMenu() {
        buttonSettings.visible()
        buttonSettings.setOnClickListener {
            EnvironmentSwitcher(this, PrefsUtil(this)).showDebugMenu()
        }
    }

    override fun showToast(message: String, toastType: String) = toast(message, toastType)

    private fun startCreateActivity() {
        val intent = Intent(this, CreateWalletActivity::class.java)
        startActivity(intent)
    }

    private fun startLoginActivity() {
        val intent = Intent(this, LoginActivity::class.java)
        startActivity(intent)
    }

    private fun startRecoveryActivityFlow() {
        val intent = Intent(this, RecoverFundsActivity::class.java)
        startActivity(intent)
    }

    private fun showFundRecoveryWarning() {
        AlertDialog.Builder(this, R.style.AlertDialogStyle)
                .setTitle(R.string.app_name)
                .setMessage(R.string.recover_funds_warning_message)
                .setPositiveButton(R.string.dialog_continue) { _, _ -> startRecoveryActivityFlow() }
                .setNegativeButton(android.R.string.cancel, null)
                .create()
                .show()
    }

    private fun showConnectivityWarning() {
        AlertDialog.Builder(this, R.style.AlertDialogStyle)
                .setMessage(getString(R.string.check_connectivity_exit))
                .setCancelable(false)
                .setNegativeButton(R.string.exit) { _, _ -> finishAffinity() }
                .setPositiveButton(R.string.retry) { _, _ ->
                    val intent = Intent(this, LandingActivity::class.java)
                    intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_NEW_TASK)
                    startActivity(intent)
                }
                .create()
                .show()
    }

    override fun showWarningPrompt(alertDialog: AlertDialog) {
        val handler = Handler()
        handler.postDelayed({
            if (!isFinishing) {
                alertDialog.show()
            }
        }, 100)
    }

    companion object {

        @JvmStatic
        fun start(context: Context) {
            val intent = Intent(context, LandingActivity::class.java)
            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_NEW_TASK)
            context.startActivity(intent)
        }

    }
}
